xmodule/docs/decisions/0004-evaluate-wasm-safe-exec.rst

```diff
@@ -0,0 +1,107 @@
+   * Python 3.11 is the first version of Python that officially supports compiling to ``wasm32-wasi``, which is why it's the one we'll start with.
+   * The Python runtime must be able to accept a course's ``python_lib.zip`` file.
+   * Parity with codejail is not feasible at this time because only a handful of our popular libraries are available. However, we should include numpy (`using Fermyon's instructions <https://www.fermyon.com/blog/introducing-componentize-py>`_) and whatever pure Python libraries are practical to do so.
+   * Preamble code (e.g. certain common library imports) will be the responsibility of the runtime environments.
+   * For this first iteration, the Wasm runtime binary build process can simply be documented, and does not need to be part of any package or CI. In particular, we do *not* want to add 100 MB+ ``.wasm`` files as a direct dependency of edx-platform.
+
+#. We will add this functionality as an opt-in, experimental feature to edx-platform.
+
+   * There will be a ``CourseWaffleFlag`` toggling the feature as a whole.
+   * The ``<script>`` tags used by the ProblemBlock will be extended with two new optional attributes:
+
+     * ``runtime`` will specify the name of the runtime that the problem execution will be dispatched to.
+     * ``options`` will allow authors to enter runtime-specific directives, like asking the runtime to use Python 2.x randomization behavior to preserve compatibility.
+
+   * If there is no value for ``runtime`` OR if the feature is not enabled, the problem will use codejail as usual.
+
+#. We will keep the implementation as simple as possible.
+
+   * We will measure the latency and memory aspects of our simple version.
+   * We may eventually be able to improve startup times by using `Wizer <https://github.com/bytecodealliance/wizer>_` or forking runtime containers after they've run their preamble, but those sorts of optimizations are out of scope for this initial effort.
+
+It is likely that we will learn a lot from our initial implementation attempts, and that we will have to adjust our approach before we arrive at something that is broadly usable and supportable for the platform over the long term.
+
+Rejected Alternatives
+*********************
+
+These rejected alternatives are things we may very well do eventually–they're just out of scope for the initial implementation/evaluation.
+
+Deprecating/Replacing Codejail
+  Critical third party libraries like SciPy are not currently built for ``wasm32-wasi``, meaning that it's not practical for us to make a fully compatible replacement for codejail at the moment.
+
+Implementing as a Service
+  It's possible to do this evaluation by implementing the code sandboxing functionality in a separate service (as the `codejail service <https://github.com/eduNEXT/codejailservice>`_ does). This library approach was chosen for a few reasons:
```

Comment:
If it comes to running it as a service, Does the WASM environment work the same as codejail for Docker and Kubernetes?

Reply:
A web service that uses WASM should behave exactly the same as any other Open edX web service from the Docker/k8s point of view. And be easier than codejail from the macOS support point of view.